### ISSUE TITLE
electrumx won't shut down

### ISSUE THREAD

**Issue by dooglus** [kyuupichan/electrumx#843]
I'm running this commit: commit 94322f4583c66a65dd233623d11b15539f9da4b8 (HEAD -> master, origin/master, origin/HEAD)
on Linux.

Haven't had a problem for months, but now it is refusing to shut down.

Here's the end of the log:

    Jun 21 HH:26:10 hostname electrumx_server[951]: INFO:DB:flush #3,561 took 0.0s.  Height 581,789 txs: 426,828,163 (+1,715)
    Jun 21 HH:26:10 hostname electrumx_server[951]: INFO:BlockProcessor:processed 1 block size 1.38 MB in 7.2s
    Jun 21 HH:26:30 hostname electrumx_server[951]: INFO:MemPool:0 txs 0.00 MB touching 0 addresses
    Jun 21 HH:27:33 hostname electrumx_server[951]: INFO:MemPool:0 txs 0.00 MB touching 0 addresses
    Jun 21 HH:27:33 hostname electrumx_server[951]: INFO:MemPool:compact fee histogram: []
    Jun 21 HH:28:25 hostname systemd[1]: Stopping Electrumx...
    Jun 21 HH:28:25 hostname electrumx_server[951]: WARNING:Controller:received SIGTERM signal, initiating shutdown
    Jun 21 HH:28:25 hostname electrumx_server[951]: INFO:Controller:shutting down
    Jun 21 HH:28:25 hostname electrumx_server[951]: ERROR:Prefetcher:unexpected exception
    Jun 21 HH:28:25 hostname electrumx_server[951]: Traceback (most recent call last):
    Jun 21 HH:28:25 hostname electrumx_server[951]:   File "/home/username/electrumx/electrumx/server/block_processor.py", line 55, in main_loop
    Jun 21 HH:28:25 hostname electrumx_server[951]:     await asyncio.sleep(self.polling_delay)
    Jun 21 HH:28:25 hostname electrumx_server[951]:   File "/usr/lib/python3.7/asyncio/tasks.py", line 568, in sleep
    Jun 21 HH:28:25 hostname electrumx_server[951]:     return await future
    Jun 21 HH:28:25 hostname electrumx_server[951]: concurrent.futures._base.CancelledError
    Jun 21 HH:28:25 hostname electrumx_server[951]: ERROR:PeerManager:could not retrieve blacklist from https://electrum.org/blacklist.json:
    Jun 21 HH:28:25 hostname electrumx_server[951]: INFO:BlockProcessor:flushing to DB for a clean shutdown...
    Jun 21 HH:38:25 hostname electrumx_server[951]: ERROR:PeerManager:could not retrieve blacklist from https://electrum.org/blacklist.json: Cannot connect to host electrum.org:443 ssl:None [Name or service not known]
    Jun 21 HH:39:30 hostname electrumx_server[951]: WARNING:Controller:received SIGTERM signal, initiating shutdown
    Jun 21 HH:44:12 hostname electrumx_server[951]: WARNING:Controller:received SIGINT signal, initiating shutdown
    Jun 21 HH:48:32 hostname electrumx_server[951]: INFO:PeerManager:blacklist from https://electrum.org/blacklist.json has 544 entries

I used `sudo systemctl stop electrumx` to try to shut it down. Then when that didn't work I tried `kill` and `kill -INT`. You see the result of each in the log, but still the process continues.

No files have recent modification dates, so I don't think it's really doing anything.

Is it safe to shut down the machine? I don't want to have to start reindexing the chain from scratch.

**Comment by streetcrypto7** [comment]
Same here, wouldn't shut down for more than 30 minutes.

**Comment by dooglus** [comment]
I finished up having to use a `kill -9` to stop it. It restarted fine after that and I haven't had a problem since.
The one time it got stuck shutting down I remember my computer was very busy when I requested the shutdown. I think I was building bitcoin-qt on all CPU cores at the time, so maybe that exacerbated a pre-existing race condition in the shut-down code causing things to get deadlocked.
Maybe. lol
